tests/python/gaia-ui-tests/gaiatest/apps/homescreen/app.py

```diff
@@ -172,14 +110,45 @@ def number_of_columns(self):
         Wait(self.marionette).until(lambda m: element.get_attribute('cols') is not None)
         return int(element.get_attribute('cols'))
 
-    class InstalledApp(PageRegion):
+    def delete_app(self, app_manifest):
+        app = self.installed_app(app_manifest)
+        remove_action_element = self.marionette.find_element(*self._remove_locator)
+        app.move_to(remove_action_element)
+
+        from gaiatest.apps.homescreen.regions.confirm_dialog import ConfirmDialog
+        return ConfirmDialog(self.marionette)
+
+    class HomescreenElement(PageRegion):
 
         @property
         def manifest_url(self):
             return self.root_element.get_attribute('data-identifier')
 
+        @property
+        def is_app_installed(self):
+            self.marionette.switch_to_shadow_root(self.root_element)
```

Comment:
Yeah, that'd be a good practice. There's also a way to prevent one person to forget about the shadow_root switching. We can define a python decorator, this way: 
```py
@WebComponent
class MyWebComponent(Widget):
    def do_something_inside_the_shadow_dom(self):
         # no call to switch_to_shadow_root at all
         self.marionette.find_element(...)
         # and so on
```
and the decorator would be defined, likely this way:
```py
class WebComponent(decorator):
    # for each function of the class decorated:
    #    before the function is called => switch to the root_element
    #    after => switch back to the original context
```

I'm not sure about the implementation details yet, but I know this can be done in Python. I filed [bug 1217826](https://bugzilla.mozilla.org/show_bug.cgi?id=1217826) to track that.